### ISSUE TITLE
No validation warnings despite very obviously incorrect synchronization

### ISSUE THREAD

**Issue by nicebyte** [KhronosGroup/Vulkan-ValidationLayers#6061]
**Environment:**
 - OS:  Windows 10
 - GPU: NVIDIA GeForce RTX 2080 Super
 - SDK or header version if building from repo:  1.3.243.0
 - Options enabled (synchronization, best practices, etc.): Core, Synchronization,  QueueSubmit Synchronization, Thread Safety, Handle Wrapping, Object Lifetime, Stateless Parameter

**Describe the Issue**

A single compute dispatch writes to a buffer.
A render pass then uses that buffer as a source of vertex data. It reads from the same buffer segment that the compute shader writes into.
There is no synchronization between the dispatch and the render pass at all.
Since theoretically the compute and graphics work may overlap, this should be a pretty obvious case of a conflict.
However, validation layers produce no warnings.

I observe the same behavior when the compute dispatch writes to an image, which is later used as a texture inside a render pass (the image uses VK_IMAGE_LAYOUT_GENERAL throughout both the compute and render pass) - validation layers produce no warnings.

**Expected behavior**

I would expect some sort of diagnostic message warning that there is a potential conflict here.

**Additional context**

When discussing this issue, it had been pointed out to me that it would be impossible to produce a useful warning here without analyzing what the shader is actually doing - it's possible to bind a storage buffer or image and never write to them, such cases shouldn't produce a warning. 

This explanation does make sense (even though it makes me much less confident in my vulkan code despite it having no validation errors).  However, another engineer called out the fact that having no barrier between a compute dispatch producing indirect draws and the actual indirect draw call used to emit warnings, but now also  does not. Therefore, it is possible that something had regressed within the validation layers.

I would like to get some clarity on this matter.

Is this the expected behavior for the validation layers? 

If yes, please treat this issue as a request to add such diagnostics. I understand this may be challenging, but I would even settle for a conservative configurator option like  "assume the content of bound storage images / buffers is always modified".  It would be better than potentially writing synchronization bugs!

**Comment by artem-lunarg** [comment]
@nicebyte thanks a lot for the detailed report and suggestions.

I'll structure the answer as a sequence of statements to better highlight each one separately.

* 💔 Yes, synchronization validation of compute writes currently is not supported
* 💔 It's non-trivial to implement. Full solution requires shader instrumentation and analysis at runtime
* 💔 This task can be of lower priority until we have open issues with more basic usage of the synchronization apis (the work is ongoing!)
* 💔/😊 We can't make the most simple approach, which detects if the shader has write instructions, to be part of the main validation since it can produce false-positives (due to branches). Especially for the apps that have large uber shaders or that use conditions to minimize the number of shader permutations. One option is to have this as a _step zero_ and make this part of the best practices validation. Still depends on the amount of false-positives and I would rather start with a step one (see below)
* 😊 Writes can be conditional that's why it's hard/impossible to rely only on static analysis. One conservative solution that uses only static analysis and does not produce false-positives (but will miss conditional writes) is to detect write instructions _that are outside of conditional blocks_. This can be the _first step_
* 😊 We have the ongoing work with GPU-AV to track memory accesses. Potentially the results from that work can be useful to implement full solution with precise access tracking
* 😊 I agree it's an important area since the compute share is only increasing, and you can't underestimate how positively this will affect application correctness/robustness. The plan is to support this in some form

I don't have an answer for a special case with the writes used by an indirect call. I will answer here if I have more information.

**Comment by nicebyte** [comment]
Thanks for the response!

All of this makes sense. I would like to offer some thoughts.

First, I totally understand why the conservative option (detecting whether there are any writes at all) can't be part of the default validation. After all, if the write depends on e.g. a uniform value, and the app _knows_ the value, it can omit that barrier safely. It's totally valid behavior and should not emit a validation error.

However, a lot of applications do not manage their barriers on such "microscopic" level, especially if they use Vulkan indirectly, through some sort of an abstraction layer. The code emitting the barriers is often itself totally unaware of details that may affect the behavior of the shader.

Which brings me to the main point: I think, in my situation, what you described as _step zero_ might actually be more useful than _step one_. The reason is, i would like to be aware that a barrier may be necessary regardless of the writes being conditional. In my case, I've no direct control over the shaders and no visibility into their behavior -- the barriers have to be conservative.

I do feel, however, that both of these approaches can co-exist - if _step one_ is implemented then _step zero_ should hopefully be trivial, and there may be a configurator option to choose the type of behavior that makes sense for the application under test.

**Comment by jeremyong** [comment]
FWIW, I think validation should _never_ emit a false positive, and it's not worth providing a validation mode that flags a hazard that doesn't actually occur (due to conditional branching as mentioned before).

Perhaps you could have a warning that triggers this, but I suspect it would be noisy, and likely not very useful in the general case, but possibly useful to enumerate candidate spots where a bug may occur.

> The code emitting the barriers is often itself totally unaware of details that may affect the behavior of the shader.

Anecdotally, there are engines/frameworks for which this isn't actually true. Graphics engineers describe what state transitions are needed, or material libraries that produce shaders from material graphs inform the system what barriers are needed.

**Comment by nicebyte** [comment]
> FWIW, I think validation should never emit a false positive

I agree with that inasmuch as validation _errors_ must all provide the specific VU clause that is being violated (and none are being violated by omitting a barrier that is provably not needed).

>  and it's not worth providing a validation mode that flags a hazard that doesn't actually occur

I would agree with this if we lived in a perfect world where detailed hazard tracking mentioned above already existed. There would simply be no need for such validation mode. Unfortunately we're not there yet, and anything is an improvement over the status quo. I'll gladly deal with the false positives if it means my code can be made more reliable.

> Graphics engineers describe what state transitions are needed

That is precisely the situation I am trying to avoid for reasons that are beyond the scope of this issue :-)

**Comment by jeremyong** [comment]
> That is precisely the situation I am trying to avoid for reasons that are beyond the scope of this issue :-)

I'm not sure this is possible currently, given that the graphics programmer that authored the (e.g.) compute shader understands under what conditions a buffer is written to, in which case, the same graphics programmer needs to describe the barrier access/scope appropriately. There's no "general" substitute for this, so if your framework wanted to automate this somehow, it would need to do so in user space at the moment.

**Comment by artem-lunarg** [comment]
Thanks for the above discussion. I think there is a good way to minimize compromises and provide great value for the end user.

**Comments**

> FWIW, I think validation should never emit a false positive

Yes, no plans for false positives. Validation should be as transparent as possible (pencil experience). In the ideal world, validation is completely transparent, and there are no reasons not to run it always. In the real world, we try to minimize the difference, especially the aspects that force the user to keep the validation turned off even if they prefer not to.

> Which brings me to the main point: I think, in my situation, what you described as step zero might actually be more useful than step one.

Yes! It's common to add a barrier conservatively  without a _perfect_ analysis of the shader code paths. This can benefit a lot of users. There's a chance step zero can be implemented in a simple, robust, and performant way (more details below). There is also a chance we miss the aspects that minimize the usefulness of this approach.

> I do feel, however, that both of these approaches can co-exist

+1

**More details**

Both approaches for placing barriers have their place and are used in the industry in applications of different sizes and complexity.

1. _Place a barrier if there is a chance the shader will perform a write operation_. That's a great point from @nicebyte. It's not always possible to determine current configuration of the shader - it could be a black box, and the engine does not provide more details. It can be possible to do this, but it's hard to analyze and could be error-prone. A simple solution might be preferable.

2. _Do your best to avoid a barrier (perfect barrier placement)_.  In the core parts of the engine or the performance-critical paths, developers might be willing to do more complex analysis and maintain `code<->shader` dependency for performance reasons. Some applications can have conventions on how the shaders are used (or dedicated systems), which simplifies (or even automates) conditional barrier placement. In a simple app (demo/sample), it might be easy to do conditional barrier placement due to app simplicity and if you want to do this, this should not produce false positives.

**Suggestion**

_Step zero_ (report that shader _can_ write but is not followed by any barrier)  is what we might want from the static part of the solution. It benefits  a lot of users, relatively simple to implement and maintain, should be performant enough to keep it always on.

As noted above, this can't be part of the standard validation due to false positives but should be enabled with an additional option. We should not treat this as the "best practice" (although it could be part of the best practices module) or as potential errors because both approaches to placing barriers are valid, and it's up to the app to decide. We can call this feature a _validation request_ , _reporting request_, etc. Such a request does not report spec violations or potential errors but provides the information we requested. In this case, it will be like, "Please report all places where the shader has write instruction (so it can potentially write), but there is no barrier before the next access." This request reports all such findings without saying it's good or bad.

I no longer see the point of _step one_, which does more complex static analysis. It might result in more noise when during shader modification, you have ping-pong requests either to add a barrier or to remove it. For simplicity, it's better to have simple _step zero_, which covers a lot of use cases and has nice properties. And if you want more, it's the responsibility of a full-fledged solution with runtime validation, potentially measurable performance hit, and complex implementation. It might take a non-trivial amount of time to implement, but it provides precise results.

**Plan**:
* Add an option for the _validation request_ that reports all places where a shader that potentially can perform write operations is not followed by any barrier. This can happen sooner.
* Implement shader instrumentation or another way to track shader execution to precisely detect write events and perform precise hazard validation, as this is done in other parts of the syncval module. This can take a lot of time.

**Comment by nicebyte** [comment]
That plan sounds good to me. Thank you very much!
